File: Medium/200. Number of Islands/solution.java
class Solution {
    int[][] di = {{-1, 0}, {1,0},{0,1},{0,-1}};
    int len, width;
    public int numIslands(char[][] grid) {
        len = grid.length;
        if(len == 0) {
            return 0;
        }
        width = grid[0].length;
        boolean[][] visited = new boolean[len][width];
        for(int i = 0; i<len;i++) {
            for(int j = 0; j<width;j++) {
                visited[i][j] = false;
            }
        }
        int ans = 0;
        for(int i = 0; i<len;i++) {
            for(int j = 0; j< width;j++) {
                if(grid[i][j] == '1' && !visited[i][j]) {
                    ans++;
                    dfs(grid, i, j, visited);
                } 
            }
        }
        return ans;
    }
    private void dfs(char[][] grid, int x, int y, boolean[][] visited) {
        visited[x][y] = true;
        for(int i = 0; i< 4;i++) {
            int newx = x + di[i][0];
            int newy = y + di[i][1];
            if(inArea(newx, newy)) {
                if(grid[newx][newy] == '1' && !visited[newx][newy]) {
                    dfs(grid, newx, newy, visited);
                }
            }
        }
        return;
    }
    private boolean inArea(int x, int y) {
        return x >= 0 && x < len && y >= 0 && y < width;
    }
}

File: Medium/200. Number of Islands/solution.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        
        if not grid or len(grid) == 0 or len(grid[0]) == 0:
            return 0
        
        rows = len(grid)
        cols = len(grid[0])
        count = 0
        for i in range(rows):
            for j in range(cols):
                if grid[i][j] == "1":
                    self.dfs(grid, i, j)
                    count += 1
        return count
                    
    
    def dfs(self, grid: List[List[str]], x: int, y: int) -> None:
        
        
        grid[x][y] = "0"
        rows = len(grid)
        cols = len(grid[0])
        dirs = [[1, 0], [0, 1], [-1, 0], [0, -1]]
        for dir in dirs:
            newx = x + dir[0]
            newy = y + dir[1]
            if newx >= 0 and newx < rows and newy >= 0 and newy < cols:
                if grid[newx][newy] == "1":
                    self.dfs(grid, newx, newy)
        return
        